feat: create `no-deprecated-functions` (#560)
jest-community/eslint-plugin-jest: docs/rules/no-deprecated-functions.md

# Warns on usage of deprecated functions (no-deprecated-functions)

Over the years Jest has accrued some debt in the form of functions that have
either been renamed for clarity, or replaced with more powerful APIs.

While typically these deprecated functions are kept in the codebase for a number
of majors, eventually they are removed completely.

## Rule details

This rule warns about calls to deprecated functions, and provides details on
what to replace them with.

This rule can also autofix a number of these deprecations for you.

### `require.requireActual` & `require.requireMock`

These functions were replaced in Jest 21 and removed in Jest 26.

Originally, the `requireActual` & `requireMock` the `requireActual`&
`requireMock` functions were placed onto the `require` function.

These functions were later moved onto the `jest` object in order to be easier
for type checkers to handle, and their use via `require` deprecated. Finally,
the release of Jest 26 saw them removed from the `require` function all
together.

### `jest.addMatchers`

> This function has been replaced with `expect.extend`, and is scheduled for

Reviewer: ```suggestion
This function was replaced with `expect.extend` in Jest 17, and is scheduled for
```

I think 17? https://github.com/facebook/jest/pull/1933

removal in Jest 27.

### `jest.resetModuleRegistry`

> This function has been renamed to `resetModules`, and is scheduled for removal

Reviewer: ```suggestion
This function was renamed to `resetModules` in Jest 15, and is scheduled for removal
```

15 😱 https://github.com/facebook/jest/pull/1511

> in Jest 26.

Reviewer: ```suggestion
in Jest 27.
```

### `jest.runTimersToTime`

> This function has been renamed to `advanceTimersByTime`, and is scheduled for

Reviewer: ```suggestion
This function was renamed to `advanceTimersByTime` in Jest 22, and is scheduled for
```

> removal in Jest 26.

Reviewer: ```suggestion
removal in Jest 27.
```

# Warns on usage of deprecated functions (no-deprecated-functions)

> Over the years jest has accrued some debt in the form of functions that have

Reviewer: ```suggestion
Over the years Jest has accrued some debt in the form of functions that have
```

either been renamed for clarity, or replaced with more powerful APIs.

While typically these deprecated functions are kept in the codebase for a number
of majors, eventually they are removed completely.

## Rule details

This rule warns about calls to deprecated functions, and provides details on
what to replace them with.

This rule can also autofix a number of these deprecations for you.

> ### `require.requireActual.` & `require.requireMock`

Reviewer: ```suggestion
### `require.requireActual` & `require.requireMock`
```

> These functions were removed in Jest 26.

Reviewer: ```suggestion
These functions were replaced in Jest 21 and removed in Jest 26.
```

> Originally in the early days of jest, the `requireActual`& `requireMock`

Reviewer: ```suggestion
Originally, the `requireActual` & `requireMock`
```

functions were placed onto the `require` function.

> These functions were later moved onto the `jest` global, and their use via

Reviewer: ```suggestion
These functions were later moved onto the `jest` object in order to be easier for type checkers to handle, and their use via
```

`require` deprecated. Finally, the release of Jest 26 saw them removed from the
`require` function all together.

> The PR implementing the removal can be found
> [here](https://github.com/facebook/jest/pull/9854).

Reviewer: ```suggestion
```

I don't think that's really relevant

### `jest.addMatchers`

> This function has been replaced with `expect.extend`, and will ideally be

Reviewer: main issue with `addMatchers` is that it's just a proxy to `jasmine.addMatchers`, and we want people to use `expect` and `expect.extend`, like you mention, directly.

It will throw an error if using `jest-circus`

Author: Might as well keep things simple if we're replacing `jasmine` with `jest-circus` 🤷 

I assume that you can call `jasmine.addMatchers` directly if you were really wanted to?

Reviewer: yup! (unless you're using circus)

> removed in Jest 27.

Reviewer: ```suggestion
and is scheduled for removal in Jest 27.
```

### `jest.resetModuleRegistry`

> This function has been renamed to `resetModules`, and will ideally be removed in
> Jest 27.

Reviewer: ```suggestion
This function has been renamed to `resetModules`, and is scheduled for removal in Jest 26.
```

### `jest.runTimersToTime`

> This function has been renamed to `advanceTimersByTime`, and will ideally be
> removed in Jest 27.

Reviewer: ```suggestion
This function has been renamed to `advanceTimersByTime`, and is scheduled for removal in Jest 26.
```
